### PR TITLE
Add 2 methods + finer checks on invalid input for alarms

### PR DESCRIPTION
**Pull request by dkm**
I needed to be able to use NaiveTime object for setting alarms. Also, it looks like the input checks for setting alarms were a bit too strict for some matching values.

### CODE REVIEW

**Comment by dkm** [comment]
I'll see how to add tests for corresponding changes

File: src/ds323x/alarms.rs [hunk 1]
```diff
@@ -127,6 +133,19 @@ where
         self.iface.write_data(&mut data)
     }
 
+    /// Set Alarm1 for a time every day.
```
Comment (eldruin) [inline src/ds323x/alarms.rs:136]: Could you add a description of the matching strategy (i.e. fires when h, m, s match)

File: src/ds323x/alarms.rs [hunk 2]
```diff
@@ -170,6 +189,18 @@ where
         self.iface.write_data(&mut data)
     }
 
+    /// Set Alarm2 for a time every day.
```
Comment (eldruin) [inline src/ds323x/alarms.rs:192]: Same as above, could you add a description of the matching strategy (i.e. fires when h, m match)

**Comment by dkm** [comment]
> Sounds good. Thank you for the improvements. As stated before, tests for the new functions would be need to be added as well.

I am trying to see how to add tests, but I have trouble with the mock interface used. I guess I have to add expected values for i2c/spi transactions, but I don't see where. Could you guide me a bit on what I should do to add test ? I've tried this to test the change in bad input filtering :

```
diff --git a/tests/alarms.rs b/tests/alarms.rs
index 6ebc786..acff9b3 100644
--- a/tests/alarms.rs
+++ b/tests/alarms.rs
@@ -80,6 +80,17 @@ mod alarm1 {
         },
         A1M::AllMatch
     );
+    set_invalid_alarm_test!(
+        day_invalid_min_1,
+        set_alarm1_day,
+        DayAlarm1 {
+            day: 1,
+            hour: Hours::H24(1),
+            minute: 60,
+            second: 1
+        },
+        A1M::SecondsMatch
+    );
     set_invalid_alarm_test!(
         day_invalid_h,
         set_alarm1_day,
```

But then I get
```
...
---- alarm1::day_invalid_min_1::cannot_set_invalid_ds3234 stdout ----
thread 'alarm1::day_invalid_min_1::cannot_set_invalid_ds3234' panicked at 'no expectation for spi::write call
...
```

**Comment by eldruin** [comment]
I see. The tests use macros heavily due to the amount of duplicated code otherwise, but their structure can seem complicated.

The reason for that error is that with your changes when the alarm is configured with `SecondsMatch` the invalid `minute: 60` is ignored and thus the provided values are valid. In this case, it proceeds with setting the values in the device, and thus needs an expectation for the bus transactions.
An example invalid input test would be:
```rust
set_invalid_alarm_test!(
        day_invalid_second,
        set_alarm1_day,
        DayAlarm1 {
            day: 1,
            hour: Hours::H24(1),
            minute: 59,
            second: 60
        },
        A1M::SecondsMatch
    );
```
A valid input test case would be something like this:
```rust
set_alarm_test!(
        match_s,
        set_alarm1_day,
        DayAlarm1 {
            day: 1,
            hour: Hours::H24(2),
            minute: 60,
            second: 4
        },
        A1M::SecondsMatch,
        ALARM1_SECONDS,
        [4, AM | 0, AM | 2, AM | 1]
    );
```
I just noticed that code that corrects previously invalid input values needs to be added. Otherwise necessary bits may be overwritten with wrong values.
I would say a sensible approach would be to set the values to the minimum valid (only) when an invalid value is received.
For the test example above the minute value (60) would need to be corrected to 0.
This needs to be documented in all the methods as well.

I think the code will get a bit messy with all these corrections depending on the input matching strategies. If you do not really need this and just became aware of it when writing `day: 0` in the new methods, maybe it is easier to just write `day: 1` there.

**Comment by dkm** [comment]
I've tried to see if I could come up with something that fixes your previous remarks (modifying invalid but ignored value to correct values) and have some tests. Please tell me if you find it too messy. If that's the case, then I'll do what you mentioned (remove the finer filtering).

As for the tests of the new functions, that's still not good enough.
The macros are expecting 2 arguments functions, but the new ones only take 1, so I've duplicated the macro code as a first step, and it nearly works (I still get an error when testing the Alarm2 _hm function). I'll iterate again later.

**Comment by eldruin** [comment]
The correction code seems fine. The tests look good so far, although I would rename them to try to make them a bit more descriptive (e.g. I did not understand what the `_bis` suffix is, probably not meant as latin for second, which is what first came to mind).
On other driver crates I have defined more general macros that accept a [variable number of arguments](https://github.com/eldruin/ds1307-rs/blob/aadf636dce5d59112405f0b7c18150023c096057/tests/common/mod.rs#L42) but it will be some work to adapt all tests to the new signature since the parameters must go at the end.

**Comment by dkm** [comment]
Yes, sorry, should have made the name better, I'm mostly learning here, so there's a lot of trial and error. I could see how to adapt the macros also if you think it's a good idea and if you don't mind having to give your feedback regularly.

**Comment by eldruin** [comment]
Sure, no problem. Glad to help onboarding rustaceans :)

**Comment by dkm** [comment]
Ok, it was easier than I expected, thanks for the pointer. I still have to change some names and add some tests for valid inputs

**Comment by eldruin** [comment]
Sounds good. You can also mark this PR as a draft while it is not ready.

**Comment by dkm** [comment]
> Sounds good. You can also mark this PR as a draft while it is not ready.

I think only reviewers can change this state, but you're right, it should be made draft :)

**Comment by eldruin** [comment]
True, thanks. I marked it as a draft now.

**Comment by dkm** [comment]
While trying to have all `if`s covered by tests, I've realized that there's an asymmetry in inputs, as hours are not handled like the others fields. But having the same logic of «incorrect but ignored input is set to sensible value» is not straightforward for `hour`. Is it a problem ? Apart from that, I think it should be ok... Hope so, at least.

File: src/ds323x/alarms.rs [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 use super::super::{BitFlags, Ds323x, Error, Hours, Register};
 use super::{decimal_to_packed_bcd, hours_to_register};
+use crate::ds323x::{NaiveTime, Timelike};
 use interface::{ReadData, WriteData};
 
 /// Parameters for setting Alarm1 on a day of the month
```
Comment (eldruin) [inline src/ds323x/alarms.rs:8]: Could you document here how the values of invalid inputs are adapted when these are ignored by the selected matching  strategy?

File: src/ds323x/alarms.rs [hunk 4]
```diff
@@ -113,20 +114,46 @@ where
...
             hours_to_register(when.hour)? | match_mask[2],
-            decimal_to_packed_bcd(when.day) | match_mask[3],
+            decimal_to_packed_bcd(day) | match_mask[3],
         ];
         self.iface.write_data(&mut data)
     }
 
+    /// Set Alarm1 for a time (fires when hours, minutes and seconds match).
+    ///
+    /// Will return an `Error::InvalidInputData` if any of the parameters is out of range.
```
Comment (eldruin) [inline src/ds323x/alarms.rs:146]: Could you document that since the day it is not used it is set to 1?

File: src/ds323x/alarms.rs [hunk 5]
```diff
@@ -113,20 +114,46 @@ where
...
+            hour: Hours::H24(when.hour() as u8),
+            minute: when.minute() as u8,
+            second: when.second() as u8,
+        };
+        self.set_alarm1_day(alarm, Alarm1Matching::HoursMinutesAndSecondsMatch)
+    }
+
     /// Set Alarm1 for weekday.
     ///
     /// Will return an `Error::InvalidInputData` if any of the parameters is out of range.
```
Comment (eldruin) [inline src/ds323x/alarms.rs:159]: Same as above: Could you document here how the values of invalid inputs are adapted when these are ignored by the selected matching strategy?

File: src/ds323x/alarms.rs [hunk 6]
```diff
@@ -157,19 +197,40 @@ where
         when: DayAlarm2,
         matching: Alarm2Matching,
     ) -> Result<(), Error<CommE, PinE>> {
```
Comment (eldruin) [inline src/ds323x/alarms.rs:199]: Same as above: Could you document here how the values of invalid inputs are adapted when these are ignored by the selected matching strategy?

File: src/ds323x/alarms.rs [hunk 7]
```diff
@@ -157,19 +197,40 @@ where
...
             hours_to_register(when.hour)? | match_mask[1],
-            decimal_to_packed_bcd(when.day) | match_mask[2],
+            decimal_to_packed_bcd(day) | match_mask[2],
         ];
         self.iface.write_data(&mut data)
     }
 
+    /// Set Alarm2 for a time (fires when hours, minutes and seconds match).
+    ///
+    /// Will return an `Error::InvalidInputData` if any of the parameters is out of range.
```
Comment (eldruin) [inline src/ds323x/alarms.rs:224]: Could you please also document here that the day will be set to 1?

File: src/ds323x/alarms.rs [hunk 8]
```diff
@@ -178,15 +239,23 @@ where
         when: WeekdayAlarm2,
         matching: Alarm2Matching,
     ) -> Result<(), Error<CommE, PinE>> {
-        if when.weekday < 1 || when.weekday > 7 || when.minute > 59 {
+        let weekday_invalid = when.weekday < 1 || when.weekday > 7;
```
Comment (eldruin) [inline src/ds323x/alarms.rs:242]: Same as above: Could you document here how the values of invalid inputs are adapted when these are ignored by the selected matching strategy?

File: tests/alarms.rs [hunk 9]
```diff
@@ -486,94 +556,140 @@ mod alarm1_day {
...
             day: 1,
             hour: Hours::H24(2),
             minute: 3,
             second: 4
         },
-        A1M::SecondsMatch,
+        A1M::SecondsMatch
+    );
+    set_alarm_test!(
+        match_s_ign_incorrect_min,
```
Comment (eldruin) [inline tests/alarms.rs:669]: Could you please replace "`_ign_`" with "`_ignore_`" in the tests? That word seems to important to be missed.

**Comment by dkm** [comment]
Hope to have all remarks handled. Feel free to request for more changes if needed !

**Comment by eldruin** [comment]
Everything is great. Thank you for your work! I will comment here when I release this.

**Comment by eldruin** [comment]
For the record, I have published this in version 0.3.1.

**Comment by dkm** [comment]
Thanks for taking the time to review my changes, and great news for the new release !

**Pull request merged**